xmtp_mls/src/groups/commit_log.rs

```diff
@@ -344,46 +316,248 @@ where
     }
-}
 
-#[cfg(test)]
-pub enum CommitLogTestFunction {
-    PublishCommitLogsToRemote,
-    SaveRemoteCommitLog,
-    All,
+    fn should_skip_remote_commit_log_entry(
+        &self,
+        validation_info: &RemoteLogValidationInfo,
+        entry: &PlaintextCommitLogEntry,
+    ) -> bool {
+        let is_applied = entry.commit_result == ProtoCommitResult::Applied as i32;
+        // Should skip if:
+        // 1. The entry signature is invalid - TODO(cam)
+        // 2. The group_id of the entry does not match the requested group_id.
+        // 3. The commit_sequence_id of the entry is <= 0.
+        // 4. The commit_sequence_id of the entry is not greater than the most recently stored entry, if one exists.
+        // 5. The last_epoch_authenticator does not match the epoch_authenticatorof the most recently stored entry with a CommitResult of COMMIT_RESULT_APPLIED, if one exists.
+        // 7. The entry has a CommitResult of COMMIT_RESULT_APPLIED, but the epoch number is not exactly 1 greater than the most recently stored entry with a result of COMMIT_RESULT_APPLIED, if one exists.
+        entry.group_id != validation_info.requested_group_id
+            || entry.commit_sequence_id == 0
+            || entry.commit_sequence_id <= validation_info.latest_stored_commit_sequence_id
+            || (is_applied
+                && !validation_info
+                    .latest_applied_epoch_authenticator
+                    .is_empty()
+                && entry.last_epoch_authenticator
+                    != validation_info.latest_applied_epoch_authenticator)
+            || (is_applied
+                && entry.applied_epoch_number != validation_info.latest_applied_epoch_number + 1)
+    }
+
+    pub async fn check_forked_state(
+        &mut self,
+    ) -> Result<HashMap<Vec<u8>, ForkedStateCheckResult>, CommitLogError> {
+        let conn = &self.context.db();
+        let conversation_ids_for_forked_state_check =
+            conn.get_conversation_ids_for_remote_log_download()?;
+
+        let mut forked_state_check_results = HashMap::new();
+
+        for conversation_id in conversation_ids_for_forked_state_check {
+            let fork_result = self.check_conversation_fork_state(conn, &conversation_id)?;
+            forked_state_check_results.insert(conversation_id, fork_result);
+        }
+
+        Ok(forked_state_check_results)
+    }
+
+    fn check_conversation_fork_state(
+        &self,
+        conn: &impl DbQuery,
+        conversation_id: &[u8],
+    ) -> Result<ForkedStateCheckResult, CommitLogError> {
+        // Get cursors for this conversation
+        let fork_check_local_cursor = conn.get_last_cursor_for_id(
+            conversation_id,
+            xmtp_db::refresh_state::EntityKind::CommitLogForkCheckLocal,
+        )?;
+        let fork_check_remote_cursor = conn.get_last_cursor_for_id(
+            conversation_id,
+            xmtp_db::refresh_state::EntityKind::CommitLogForkCheckRemote,
+        )?;
```

Comment:
Not important for this PR, just something I thought about with the consensus key stuff too. Would it be fair to say that we expect no concurrency with any of this code? Wondering if it makes sense to prove it by acquiring a lock on each iteration of the worker (which self-documents the concurrency model and protects against any unanticipated scenarios, like the previous iteration is still running when the next iteration starts, or there are two workers or something)

Reply:
ill make a note to revisit this for all of our workers before we turn on auto fork recovery 👍